json-decode.go

```diff
@@ -61,6 +61,29 @@ func (cdc *Codec) decodeReflectJSON(bz []byte, info *TypeInfo, rv reflect.Value,
 		}
 	}
 
+	// Handle override if a pointer to rv implements UnmarshalAminoJSON.
+	if info.IsAminoUnmarshalerJSON {
+		// First, decode repr instance from JSON.
+		rrv := reflect.New(info.AminoUnmarshalJSONReprType).Elem()
+		var rinfo *TypeInfo
+		rinfo, err = cdc.getTypeInfoWlock(info.AminoUnmarshalJSONReprType)
+		if err != nil {
+			return
+		}
+		err = cdc.decodeReflectJSON(bz, rinfo, rrv, fopts)
+		if err != nil {
+			return
+		}
+		// Then, decode from repr instance.
+		uwrm := rv.Addr().MethodByName("UnmarshalAminoJSON")
+		uwouts := uwrm.Call([]reflect.Value{rrv})
+		erri := uwouts[0].Interface()
+		if erri != nil {
+			err = erri.(error)
+		}
+		return
+	}
```

Comment:
Why can't we just do:
```go
	// Handle override if a pointer to rv implements json.Unmarshaler.
	if rv.Addr().Type().Implements(jsonUnmarshalerType) {
		err = rv.Addr().Interface().(AminoJSONUnmarshaler).UnmarshalAminoJSON(bz)
		return
	}
```

Is this because we want to allow any type to be used instead of `[]byte`? Is it better to be consistent with `Un/MarshalAmino` or should we just do the simpler thing and use the same signatures as `Un/MarshalJSON`?

Reply:
> Is it better to be consistent with Un/MarshalAmino or should we just do the simpler thing and use the same signatures as Un/MarshalJSON

Yes, that's the whole debate. The way I went for is to keep the symmetry in function signature between:
- Un/MarshalJSON and Un/MarshalBinary
- Un/MarshalAminoJSON and Un/MarshalAmino